### PR TITLE
🔴🔴  [config] Add `get_sub_config` (in place of `get_text_config`) and fix related bugs

### LINKED ISSUE #40644: `BlenderbotForConditionalGeneration` errors out with `list index out of range`
### System Info

Using Google Colab

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Colab notebook: https://colab.research.google.com/gist/ariG23498/063c2888c41afebb71ca614278161445/scratchpad.ipynb

### Expected behavior

Should work as expected. The code has been taken from the official docs: https://huggingface.co/docs/transformers/en/model_doc/blenderbot#transformers.BlenderbotForConditionalGeneration

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR (🔴 = BC breaking):
1. Adds `config.get_sub_config()`. This function is an upgrade to `get_text_config`:
  a. Allows model-agnostic extraction of different modalities in a single function, future-proofing its uses;
  b. Uses heuristics to determine correct encoder/decoder and modality matches (as opposed to relying on hardcoded sub config names that could match). These heuristics prevent us from having to manually update matching names as new models come out. They also nudge future models towards the same patterns (👉 standardization)
  c. Related to #40277: when pulling `encoder`/`decoder` attributes from older encoder-decoder configs, make sure to consider `config.attribute_map`
  d. Sometimes we want any text config (e.g resizing embeddings), other times we want any decoder modality (e.g. setting up the kv cache)... and sometimes we want specifically the text decoder. `get_text_config` didn't support this level of control, but `get_sub_config` does 🤗 
  e. (b.) + (c.) + (d.) = several bug fixes + allow us to remove a few overwrites
2. 🔴 Deprecates `get_text_config`, replacing its calls by `get_sub_config(modality="text", ...)`. This replacement is BC-breaking because of 1.b. However, in most cases, it fixes subtle bugs.
3. Enables `SlidingWindowLayer`, the static one, on cross-attention cases (where no `cache_position` is provided).
4. Corrects `is_encoder_decoder` check in `generate` tests.
5. Fixes a few related corner cases in `test_generate_continue_from_inputs_embeds` and `test_past_key_values_format` (which allows us to delete skips/overwrites)
___________________________

This PR is a follow-up to #40454 (incomplete fix for the underlying issues, which ended up being quite complex)
Fixes #40644

### CODE REVIEW

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -421,7 +421,7 @@ def test_greedy_generate(self):
             model = model_class(config).to(torch_device).eval()
             output_generate = self._greedy_generate(model=model, inputs_dict=inputs_dict)
 
-            if model.config.get_text_config(decoder=True).is_encoder_decoder:
+            if model.config.is_encoder_decoder:
```
Comment (gante) [inline tests/generation/test_utils.py:339]: The original `model.config.get_text_config(decoder=True).is_encoder_decoder` was added because of `BLIP2`, which was not setting its outer-level `is_encoder_decoder` attribute correctly. This PR fixes the root issue there.

In practice, this line was not right: if we want to check if a model is an encoder-decoder model, we can't look at it's decoder config only :)

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -1436,29 +1438,36 @@ def test_generate_continue_from_past_key_values(self):
...
     @pytest.mark.generate
     def test_generate_continue_from_inputs_embeds(self):
-        """Tests that we can continue generation from `inputs_embeds` and past key values returned from a previous `generate` call."""
+        """
+        Tests that we can continue generation from `inputs_embeds` and past key values returned from a previous
+        `generate` call.
+        """
         for model_class in self.all_generative_model_classes:
-            if any(model_name in model_class.__name__.lower() for model_name in ["imagegpt"]):
+            # To be more precise: technically we can run this test on all models that have `inputs_embeds` or
```
Comment (gante) [inline tests/generation/test_utils.py:1440]: tl;dr moved model-level overwrites here, so we can better track odd models, and updated skip conditions

File: src/transformers/cache_utils.py [hunk 3]
```diff
@@ -1009,7 +1010,9 @@ def __init__(
         layers = []
         # If a config is passed, use it to infer the layer types and initialize accordingly
         if config is not None:
-            config = config.get_text_config()
+            # We pull the decoder sub-config here to allow composite models to easily initialize the cache as
+            # `DynamicCache(config=model.config)`
+            config = config.get_sub_config(decoder=True)
```
Comment (gante) [inline src/transformers/cache_utils.py]: e.g. in text-to-speech models, the decoder is not a text decoder :) 

(it was working before because all TTS decoders had a common name in their decoder config)

**Comment by gante** [comment]
Closing, the root issue is solved in #40939 (the remaining fixes here were targetting old models)
